### DISCUSSION TITLE
How to arrange P2P instead of client-server?

### DISCUSSION THREAD

**Discussion by AArnott** [paritytech/jsonrpsee#1458] [Q&A] [answered]
In #1456 I was able to get *either* a client *or* a server set up on a duplex stream. Yay.

But the ClientBuilder seems to be oriented only around being a client. Similarly the server API (that I use in that other discussion at least) seems to be only useful for server-side.
In all of my scenarios, I use JSON-RPC as a peer-to-peer system, where either side can send and receive requests.

Is there an API to use for setting that up? Any examples?

**Comment by niklasad1** [comment, marked as answer]
No, this library doesn't expose APIs to act both as "client and server"

**Reply by AArnott** [reply]
Thanks for the reply. Is there another crate you've heard of that does? Do you foresee this crate ever offering P2P use cases? If so, would you be open to pull requests for it if I ever got around to it?
